python/graphstorm/gconstruct/transform.py

```diff
@@ -935,6 +910,149 @@ def call(self, feats):
+        self._nid_map = id_maps
+
+    def pre_process(self, feats):
+        """ Pre-process input feats
+
+            Not all the edges have the same number of hard negatives.
+            Thus we need to know the maxmun number of hard negatives first.
+
+        Parameters
+        ----------
+        feats:
+            feats to be processed
+
+        Return:
+            dict: {feature_name: feats_statistics}
+        """
+        assert isinstance(feats, (np.ndarray, ExtMemArrayWrapper)), \
+            "Feature of HardEdgeNegativeTransform must be numpy array or ExtMemArray"
+
+        if self._separator is None:
+            max_dim = feats.shape[1]
+        else:
+            assert len(feats.shape) == 1 or feats.shape[1] == 1, \
+                "When a separator is given, the input feats must be a list of strings."
+
+
+            if feats.dtype.type is not np.str_:
+                logging.warning("When a separator is given, "
+                                "the input feats must be a list of strings."
+                                "But we get %s. Will convert the input feats into str array",
+                                feats.dtype.type)
+            feats = feats.astype(str)
+            max_dim = 0
+            for feat in feats:
+                dim_size = len(feat.split(self._separator))
+                max_dim = dim_size if dim_size > max_dim else max_dim
+        return {self.feat_name: max_dim}
+
+    def update_info(self, info):
+        max_dim = max(info)
+        self._max_dim = max_dim
+
+    def call(self, feats):
+        """ Generate hard negatives as features
+
+        Parameters
+        ----------
+        feats : np array
+            Data with hard negatives.
+
+        Returns
+        -------
+        np.array
+        """
+        assert self._target_ntype is not None, \
+            "The target node type should be set, it can be the source node type " \
+            "or the destination node type depending on the hard negative case."
+        nid_map = self._nid_map[self._target_ntype]
+
+        neg_ids = np.full((len(feats), self._max_dim), -1, dtype=np.int64)
+        if self._separator is None:
+            for i, raw_ids in enumerate(feats):
+                if raw_ids is None:
+                    continue
+                # raw_ids is a numpy array
+                # handle empty strings in raw_ids
+                raw_ids = raw_ids.astype(str)
+                nid_idx = np.nonzero(raw_ids != "")[0]
+                nids, _ = nid_map.map_id(raw_ids[nid_idx].astype(
+                    nid_map.map_key_dtype))
+                neg_ids[i][nid_idx] = nids
+        else:
+            for i, feat in enumerate(feats):
+                if feat is None:
+                    continue
+                raw_ids = np.array(feat.split(self._separator))
+                nids, _ = nid_map.map_id(raw_ids.astype(nid_map.map_key_dtype))
+                neg_ids[i][:nids.shape[0]] = nids
```

Comment:
do you expect that different edges have different numbers of negative edges?

Reply:
It is possible.

Reply:
should we consider this an incorrect input?

Reply:
currently, if an edge doesn't have enough hard negative edges, it is filled with -1. we need to carefully handle -1 during the training. is it necessary to introduce the complexity?

Reply:
if there aren't enough hard negative edges, we can just duplicate them. the impact should be the same.